modules/protocol/src/vector.ts

```diff
@@ -136,26 +121,35 @@ export class Vector implements IVectorProtocol {
       channelAddress: params.channelAddress,
       updateSender: this.publicIdentifier,
     });
-    let aliceIdentifier: string;
-    let bobIdentifier: string;
-    let channel: FullChannelState | undefined;
-    if (params.type === UpdateType.setup) {
-      aliceIdentifier = this.publicIdentifier;
-      bobIdentifier = (params as UpdateParams<"setup">).details.counterpartyIdentifier;
-    } else {
-      channel = await this.storeService.getChannelState(params.channelAddress);
-      if (!channel) {
-        return Result.fail(new OutboundChannelUpdateError(OutboundChannelUpdateError.reasons.ChannelNotFound, params));
-      }
-      aliceIdentifier = channel.aliceIdentifier;
-      bobIdentifier = channel.bobIdentifier;
-    }
-    const isAlice = this.publicIdentifier === aliceIdentifier;
-    const counterpartyIdentifier = isAlice ? bobIdentifier : aliceIdentifier;
-    throw new Error("must implement internal queueing");
-    const outboundRes = await this.lockedOperation(params);
 
-    return outboundRes;
+    // Pull channel from store
+
+    // Update highest seen nonce to next nonce
+
+    // propose update using sync.outbound
+    // should:
+    // - add params to queue
```

Comment:
It's not clear to me that adding params to the queue and sending the proposal should happen in the same method.

I would instead:
Have a method for adding the params to the queue which creates a "bluebird style" manually resolvable promise. The params and the promise can be added to the queue.

Have a separate method which continuously takes works off the queue, executes it, and resolves the promise in a loop. When the queue is empty, it pushes a promise to wake itself up that is resolved when the next item is queued.

In this way, updates will be serialized as if there is a lock. There cannot be race conditions by updating stale data.

I'd be happy to write this queue for you if you think it's the right abstraction.

Reply:
The nuance here (that I punted on mostly) is when you are out of sync. 

Currently when you are out of sync you have to apply the update that you dropped, then regenerate and resend the params. This worked well in the lock-based system, but I think for the nonce based system your way of doing things is cleaner. The concern I have about the out of sync case is you have to adopt and validate your signature on an update that has a stale nonce, which may be a bit odd for the other "only higher nonces" assumption. For deposits you can mostly ignore things since it's a symmetric update, but for `create` and `resolve` updates the transfer in your store will need to be created/updated.

I think we'll have to break up the `outbound` method -- it exists as all one thing since it uses stored data to generate the `nonce` when creating the update (sent over the wire) from the params (supplied to the API).

Reply:
But yes, I think if you want to start building that that would be great. I'll be traveling all day on tomorrow, but happy to do some work on it over the weekend if you start it on a separate branch and leave some notes on where to pick it up from in the comments